xla/service/gpu/gemm_fusion_autotuner.cc

```diff
@@ -1165,11 +1166,60 @@ absl::StatusOr<bool> GemmFusionAutotuner::Run(
                                             ? "(with correctness check)"
                                             : "(without correctness check)";
 
-    VLOG(1) << "Autotuning " << gemm_config_sets.size() << " fusions "
-            << correctness_check_str << ".";
+    const bool shard_autotuning = debug_options.xla_gpu_shard_autotuning() &&
+                                  config_.shard_count() > 1 &&
+                                  total_fusion_count > 0;
+    // Trim the set of configs to what this rank has to run.
+    if (shard_autotuning) {
+      if (config_.key_value_store() == nullptr) {
+        return absl::FailedPreconditionError(
+            "Sharded autotuning requested but key-value store is missing.");
+      }
+      const int bucket_size = (total_fusion_count + config_.shard_count() - 1) /
+                              config_.shard_count();
+      const int start = bucket_size * config_.shard_index();
+      const int end = std::min(start + bucket_size, total_fusion_count);
+      if (start >= end) {
+        gemm_config_sets.clear();
+      } else {
+        gemm_config_sets = TilingConfigs(gemm_config_sets.cbegin() + start,
+                                         gemm_config_sets.cbegin() + end);
+      }
+    }
+
+    VLOG(1) << absl::StrFormat(
+        "Shard %d / %d: autotuning %d / %d fusions for %s %s.",
+        config_.shard_index() + 1, config_.shard_count(),
+        gemm_config_sets.size(), total_fusion_count, module->name(),
+        correctness_check_str);
     TF_RETURN_IF_ERROR(autotuner.Autotune(*opt_compile_util, gemm_config_sets,
                                           std::move(fusion_count_map)));
     VLOG(1) << "Done autotuning.";
+
+    // Exchange the results with the other ranks.
+    if (shard_autotuning) {
+      AutotuneResults results;
+      TF_RETURN_IF_ERROR(AutotunerUtil::SerializeAutotuneResults(&results));
+      TF_ASSIGN_OR_RETURN(std::string results_str,
+                          AutotuneResultsToString(results, true));
+      constexpr absl::string_view kKeyPrefix = "gemm_fusion_autotuning_results";
+      TF_RETURN_IF_ERROR(config_.key_value_store()->Set(
+          absl::StrFormat("%s_%d_%d", kKeyPrefix, module->unique_id(),
+                          config_.shard_index()),
+          results_str));
+      for (int i = 0; i < config_.shard_count(); ++i) {
+        if (i == config_.shard_index()) {
+          continue;
+        }
+        TF_ASSIGN_OR_RETURN(
+            std::string autotune_results_str,
+            config_.key_value_store()->Get(
+                absl::StrFormat("%s_%d_%d", kKeyPrefix, module->unique_id(), i),
+                absl::Hours(1)));
```

Comment:
Nit: any particular reason for one hour? Why remove those at all?

Reply:
It should be long to not break long compilations. Multiple minutes are easily possible; one might want to use this with exhaustive autotuning and get to tens of minutes. It shouldn't be an infinity too. What do you mean by "remove those"?

Reply:
I thought maybe it's better to stop an infinitely hanging compilation after some time. Can put infinite duration if you prefer.